### PR TITLE
Add Support for Check Constraint - Backend code changes

### PR DESCRIPTION
**Pull request by taherkl**
Backend Support for Check Constraint

Fixes #<issue_number_goes_here>

> It's a good idea to open an issue first for discussion.

- [x] Tests pass
- [ ] Appropriate changes to README are included in PR

Backend code changes to support CHECK CONSTRAINTS

1. Added DB Collation regex on the constants
2. Added helper function for generating Check constraints ID
3. Updated Conv schema to support Check Constraints
4. CRUD operation for check constraints expressions to Spanner DDL
5. Added regex pattern validation to check constraints name, table name and database collation

### CODE REVIEW

File: sources/common/toddl.go [hunk 1]
```diff
@@ -234,6 +235,21 @@ func cvtForeignKeys(conv *internal.Conv, spTableName string, srcTableId string,
...
+	for _, cks := range srcKeys {
+		spcks = append(spcks, ddl.CheckConstraint{
+			Id:   cks.Id,
+			Name: internal.ToSpannerCheckConstraintName(conv, cks.Name),
+			Expr: cks.Expr,
+		})
+
+	}
+
+	return internal.GetSpannerValidExpression(spcks)
```
Comment (asthamohta) [inline sources/common/toddl.go:250]: This API creates and deletes a spanner database. It shouldn't be called on a per table basis but rather once for all tables for the following reasons:
1. Both create and delete db are expensive and long running operations
2. Table migration is a parallelized op and there is a limit on the number of times create and delete tables can be called. Further the db created is always named the same so this can cause race conditions

File: sources/common/toddl_test.go [hunk 2]
```diff
@@ -428,3 +428,34 @@ func Test_SchemaToSpannerSequenceHelper(t *testing.T) {
 		assert.Equal(t, expectedConv, conv)
 	}
 }
+
+func Test_cvtCheckContraint(t *testing.T) {
```
Comment (asthamohta) [inline sources/common/toddl_test.go:432]: Please add a mock for "GetSpannerValidExpression" here and check results for when this doesn't return all expressions as valid

File: sources/mysql/infoschema.go [hunk 3]
```diff
@@ -237,38 +227,98 @@ func (isi InfoSchemaImpl) GetColumns(conv *internal.Conv, table common.SchemaAnd
...
-              WHERE k.TABLE_SCHEMA = ? AND k.TABLE_NAME = ? ORDER BY k.ordinal_position;`
-	rows, err := isi.Db.Query(q, table.Schema, table.Name)
+func (isi InfoSchemaImpl) GetConstraints(conv *internal.Conv, table common.SchemaAndName) ([]string, []schema.CheckConstraint, map[string][]string, error) {
+	tableExists, err := isi.isCheckConstraintsTablePresent()
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
+	}
+
+	finalQuery := isi.getQuery(tableExists)
```
Comment (asthamohta) [inline sources/mysql/infoschema.go:236]: Rename this function to getConstraintsQuery or something. This is confusing and too generic

File: sources/mysql/infoschema.go [hunk 4]
```diff
@@ -237,38 +227,98 @@ func (isi InfoSchemaImpl) GetColumns(conv *internal.Conv, table common.SchemaAnd
...
+
 	for rows.Next() {
-		err := rows.Scan(&col, &constraint)
-		if err != nil {
-			conv.Unexpected(fmt.Sprintf("Can't scan: %v", err))
-			continue
-		}
-		if col == "" || constraint == "" {
-			conv.Unexpected(fmt.Sprintf("Got empty col or constraint"))
+		if err := isi.processRow(rows, tableExists, conv, &primaryKeys, &checkKeys, m); err != nil {
```
Comment (asthamohta) [inline sources/mysql/infoschema.go:248]: Please add code to handle the error

File: spanner/ddl/ast.go [hunk 5]
```diff
@@ -494,6 +508,23 @@ func (k Foreignkey) PrintForeignKeyAlterTable(spannerSchema Schema, c Config, ta
 	return s
 }
 
+// PrintCheckConstraintTable formats the check constraints in SQL syntax.
+func PrintCheckConstraintTable(cks []CheckConstraint) string {
+	var builder strings.Builder
+
+	for _, col := range cks {
+		builder.WriteString(fmt.Sprintf("\tCONSTRAINT %s CHECK %s,\n", col.Name, col.Expr))
```
Comment (asthamohta) [inline spanner/ddl/ast.go:516]: There are check constraints possible without a name. This will fail in that case

File: spanner/ddl/ast_test.go [hunk 6]
```diff
@@ -143,6 +143,10 @@ func TestPrintCreateTable(t *testing.T) {
 			},
 			PrimaryKeys: []IndexKey{{ColId: "col1", Desc: true}},
 			ForeignKeys: nil,
+			CheckConstraints: []CheckConstraint{
+				{Id: "ck1", Name: "check_1", Expr: "(age > 18)"},
```
Comment (asthamohta) [inline spanner/ddl/ast_test.go:147]: Add test for empty Name

File: spanner/ddl/ast_test.go [hunk 7]
```diff
@@ -950,3 +989,39 @@ func TestGetSortedTableIdsBySpName(t *testing.T) {
 		})
 	}
 }
+
+func TestPrintCheckConstraintTable(t *testing.T) {
```
Comment (asthamohta) [inline spanner/ddl/ast_test.go:993]: Same as above

File: webv2/api/schema.go [hunk 8]
```diff
@@ -488,6 +488,91 @@ func RestoreSecondaryIndex(w http.ResponseWriter, r *http.Request) {
...
+
+	convm := session.ConvWithMetadata{
+		SessionMetadata: sessionState.SessionMetadata,
+		Conv:            *sessionState.Conv,
+	}
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(convm)
+}
+
+func doesNameExist(spcks []ddl.CheckConstraint, targetName string) bool {
```
Comment (asthamohta) [inline webv2/api/schema.go:525]: Add that this is for Check Constraints in the function name

File: webv2/api/schema.go [hunk 9]
```diff
@@ -488,6 +488,91 @@ func RestoreSecondaryIndex(w http.ResponseWriter, r *http.Request) {
...
+	for _, spck := range spcks {
+		if strings.Contains(spck.Expr, targetName) {
```
Comment (asthamohta) [inline webv2/api/schema.go:527]: This does a contains not exact match, can cause errors

File: webv2/api/schema.go [hunk 10]
```diff
@@ -488,6 +488,91 @@ func RestoreSecondaryIndex(w http.ResponseWriter, r *http.Request) {
...
+			if len(sp[src.Id].CheckConstraints) > 0 {
+				spType := col.T.Name
+				srcType := srcschema[src.Id].ColDefs[col.Id].Type
+				actualType := mysqlDefaultTypeMap[srcType.Name]
+				if actualType.Name != spType {
+					columnName := sp[src.Id].ColDefs[col.Id].Name
+					spcks := sp[src.Id].CheckConstraints
+					if doesNameExist(spcks, columnName) {
+						flag = false
+						schemaIssue = sessionState.Conv.SchemaIssues[src.Id].ColumnLevelIssues[col.Id]
```
Comment (asthamohta) [inline webv2/api/schema.go:560]: I don't understand why we are even allowing user to make this update. The update itself should be blocked

File: webv2/table/update_table_schema.go [hunk 11]
```diff
@@ -96,6 +98,14 @@ func UpdateTableSchema(w http.ResponseWriter, r *http.Request) {
 
 		if v.Rename != "" && v.Rename != conv.SpSchema[tableId].ColDefs[colId].Name {
 
+			oldName := conv.SrcSchema[tableId].ColDefs[colId].Name
```
Comment (asthamohta) [inline webv2/table/update_table_schema.go:101]: Same as above

**Comment by asthamohta** [comment]
There are two things missing in this PR:
1. Either a separate PR or in the same PR we need ot make setting instance and project name mandatory on the page before schema migration.
2. The validation call is not called when prepare migration is clicked.

File: internal/helpers.go [hunk 12]
```diff
@@ -65,6 +65,11 @@ func GenerateForeignkeyId() string {
 func GenerateIndexesId() string {
 	return GenerateId("i")
 }
+
+func GenerateCheckConstrainstId() string {
+	return GenerateId("ck")
```
Comment (bharadwaj-aditya) [inline internal/helpers.go:70]: nit: ck is non obvious - can we use cc or something else here ?

File: internal/reports/report_helpers.go [hunk 13]
```diff
@@ -403,6 +403,13 @@ func buildTableReportBody(conv *internal.Conv, tableId string, issues map[string
 						Description: fmt.Sprintf("UNIQUE constraint on column(s) '%s' replaced with primary key since table '%s' didn't have one. Spanner requires a primary key for every table", strings.Join(uniquePK, ", "), conv.SpSchema[tableId].Name),
 					}
 					l = append(l, toAppend)
+				case internal.TypeMismatch:
```
Comment (bharadwaj-aditya) [inline internal/reports/report_helpers.go:406]: Is this the error the expression API will return in case of all errors or type errors only ?

File: sources/mysql/infoschema.go [hunk 14]
```diff
@@ -237,38 +227,98 @@ func (isi InfoSchemaImpl) GetColumns(conv *internal.Conv, table common.SchemaAnd
...
+	if err != nil {
+		return false, err
+	}
+	return tableExistsCount > 0, nil
+}
+
+// getQuery returns the appropriate SQL query based on the existence of CHECK_CONSTRAINTS.
+func (isi InfoSchemaImpl) getQuery(tableExists bool) string {
+	if tableExists {
+		return `SELECT k.COLUMN_NAME, t.CONSTRAINT_TYPE, COALESCE(c.CHECK_CLAUSE, '') AS CHECK_CLAUSE
```
Comment (bharadwaj-aditya) [inline sources/mysql/infoschema.go:270]: do we need this fork in the logic here ? The query would return empty in case the table does not exist.

File: sources/mysql/infoschema.go [hunk 15]
```diff
@@ -237,38 +227,98 @@ func (isi InfoSchemaImpl) GetColumns(conv *internal.Conv, table common.SchemaAnd
...
+
+// processRow handles scanning and processing of a database row for GetConstraints.
+func (isi InfoSchemaImpl) processRow(
+	rows *sql.Rows, tableExists bool, conv *internal.Conv, primaryKeys *[]string,
+	checkKeys *[]schema.CheckConstraint, m map[string][]string) error {
+
+	var col, constraintType, checkClause string
+	err := rows.Scan(&col, &constraintType, &checkClause)
+	if err != nil {
+		conv.Unexpected(fmt.Sprintf("Can't scan: %v", err))
```
Comment (bharadwaj-aditya) [inline sources/mysql/infoschema.go:301]: nit: can't scan constraints. err: %v

File: sources/mysql/infoschema.go [hunk 16]
```diff
@@ -237,38 +227,98 @@ func (isi InfoSchemaImpl) GetColumns(conv *internal.Conv, table common.SchemaAnd
...
+	if col == "" || constraintType == "" {
+		conv.Unexpected("Got empty column or constraint type")
+		return nil
+	}
+
+	switch constraintType {
+	case "PRIMARY KEY":
+		*primaryKeys = append(*primaryKeys, col)
+	case "CHECK":
+		checkClause = strings.ReplaceAll(checkClause, "_utf8mb4\\", "")
```
Comment (bharadwaj-aditya) [inline sources/mysql/infoschema.go:314]: what happens in other collations ?

File: spanner/ddl/ast.go [hunk 17]
```diff
@@ -494,6 +508,23 @@ func (k Foreignkey) PrintForeignKeyAlterTable(spannerSchema Schema, c Config, ta
 	return s
 }
 
+// PrintCheckConstraintTable formats the check constraints in SQL syntax.
+func PrintCheckConstraintTable(cks []CheckConstraint) string {
```
Comment (bharadwaj-aditya) [inline spanner/ddl/ast.go:512]: nit: this is not really printing the check constraint.

Suggest this be renamed to formatCheckConstraints or something similar

File: spanner/ddl/ast.go (hunk 5 above)
Comment (bharadwaj-aditya) [reply spanner/ddl/ast.go:516]: Can you please add a test case for this ?

File: webv2/api/schema.go [hunk 18]
```diff
@@ -488,6 +488,91 @@ func RestoreSecondaryIndex(w http.ResponseWriter, r *http.Request) {
...
+	defer sessionState.Conv.ConvLock.Unlock()
+
+	sp := sessionState.Conv.SpSchema
+	srcschema := sessionState.Conv.SrcSchema
+	flag := true
+	var schemaIssue []internal.SchemaIssue
+
+	for _, src := range srcschema {
+		for _, col := range sp[src.Id].ColDefs {
+			if len(sp[src.Id].CheckConstraints) > 0 {
```
Comment (bharadwaj-aditya) [inline webv2/api/schema.go:551]: nit: 6 levels of nesting of code here. It would be good if this can be rephrased to have lesser nesting for more readable code.

**Review by bharadwaj-aditya — CHANGES_REQUESTED** [review]

File: internal/reports/report_helpers.go (hunk 13 above)
Comment (akashthawaitcc) [reply internal/reports/report_helpers.go:406]: this is for type errors only.

File: webv2/table/review_table_schema.go [hunk 19]
```diff
@@ -108,6 +108,13 @@ func ReviewTableSchema(w http.ResponseWriter, r *http.Request) {
 					return
 				}
 			}
+			oldName := conv.SrcSchema[tableId].ColDefs[colId].Name
+
+			for i := range conv.SpSchema[tableId].CheckConstraints {
+				originalString := conv.SpSchema[tableId].CheckConstraints[i].Expr
```
[the remark this answers is not in the document]
Comment (akashthawaitcc) [reply webv2/table/review_table_schema.go:113]: added regular expression to match the exact column name.

File: webv2/table/update_table_schema.go (hunk 11 above)
Comment (akashthawaitcc) [reply webv2/table/update_table_schema.go:96]: added regular expression to match the exact column name.

File: sources/mysql/infoschema.go (hunk 4 above)
Comment (akashthawaitcc) [reply sources/mysql/infoschema.go:249]: added code to handle error

File: sources/common/toddl.go (hunk 1 above)
Comment (akashthawaitcc) [reply sources/common/toddl.go]: removed this API call.

File: spanner/ddl/ast_test.go (hunk 6 above)
Comment (akashthawaitcc) [reply spanner/ddl/ast_test.go:147]: added test case

File: sources/mysql/infoschema.go (hunk 14 above)
Comment (akashthawaitcc) [reply sources/mysql/infoschema.go:272]: we tried multiple iterations of the query, it was failing on the older versions

File: spanner/ddl/ast.go (hunk 5 above)
Comment (akashthawaitcc) [reply spanner/ddl/ast.go]: Added unit test

File: sources/mysql/infoschema.go (hunk 3 above)
Comment (akashthawaitcc) [reply sources/mysql/infoschema.go]: Renamed function name to "getConstraintsDQL"

File: sources/common/toddl_test.go (hunk 2 above)
Comment (akashthawaitcc) [reply sources/common/toddl_test.go:432]: Removed "GetSpannerValidExpression" from the toddl.go, so no need to add the mock for the API call

File: webv2/api/schema.go (hunk 8 above)
Comment (akashthawaitcc) [reply webv2/api/schema.go]: Renamed functionName to "isCheckConstraintsNameExist"

File: spanner/ddl/ast_test.go (hunk 7 above)
Comment (akashthawaitcc) [reply spanner/ddl/ast_test.go]: this is resolved as above.

File: webv2/api/schema.go (hunk 9 above)
Comment (akashthawaitcc) [reply webv2/api/schema.go]: This code block is fixed.

File: webv2/api/schema.go (hunk 10 above)
Comment (akashthawaitcc) [reply webv2/api/schema.go:561]: Added fix.

File: internal/helpers.go (hunk 12 above)
Comment (akashthawaitcc) [reply internal/helpers.go]: this has been incorporated to the code.

File: sources/mysql/infoschema.go (hunk 15 above)
Comment (akashthawaitcc) [reply sources/mysql/infoschema.go]: this change has been incorporated.

File: spanner/ddl/ast.go (hunk 17 above)
Comment (akashthawaitcc) [reply spanner/ddl/ast.go]: function name renamed to "FormatCheckConstraints"

File: webv2/api/schema.go (hunk 18 above)
Comment (akashthawaitcc) [reply webv2/api/schema.go:552]: This block will be managed by verificationAPI so this code block will be removed.

**Comment by bharadwaj-aditya** [comment]
Can you please update the PR description with what is being done, and make it it a regular PR from draft ?

File: sources/mysql/infoschema.go (hunk 14 above)
Comment (bharadwaj-aditya) [reply sources/mysql/infoschema.go:272]: Can you add a comment - which versions of MySQL go in which flow ? This is not clear from the code.

File: webv2/table/review_table_schema.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (akashthawaitcc) [reply webv2/table/review_table_schema.go:113]: added test case for this flow

File: sources/mysql/infoschema.go (hunk 14 above)
Comment (bharadwaj-aditya) [reply sources/mysql/infoschema.go:272]: Bumping up this comment.

File: sources/mysql/infoschema.go [hunk 20]
```diff
@@ -237,38 +228,99 @@ func (isi InfoSchemaImpl) GetColumns(conv *internal.Conv, table common.SchemaAnd
...
-			m[col] = append(m[col], constraint)
-		}
 	}
-	return primaryKeys, m, nil
+
+	return primaryKeys, checkKeys, m, nil
+}
+
+// checkCheckConstraintsTableExists checks if the CHECK_CONSTRAINTS table exists.
+func (isi InfoSchemaImpl) isCheckConstraintsTablePresent() (bool, error) {
```
Comment (bharadwaj-aditya) [inline sources/mysql/infoschema.go:259]: nit: it seems like this method is called only from the getConstraintsDQL method. 

Why not just call it directly from within that method and avoid passing the tableExists Flag ?

File: webv2/table/review_table_schema.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (bharadwaj-aditya) [reply webv2/table/review_table_schema.go:113]: Can you point me to the test case ? I can't seem to find it here.

File: sources/mysql/infoschema.go (hunk 20 above)
Comment (akashthawaitcc) [reply sources/mysql/infoschema.go:257]: Added this fix

File: webv2/table/review_table_schema.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (akashthawaitcc) [reply webv2/table/review_table_schema.go:112]: UT is under mysql/infoschema_test.go

File: sources/mysql/infoschema.go (hunk 14 above)
Comment (akashthawaitcc) [reply sources/mysql/infoschema.go:277]: added a comment with the mysql version

File: sources/mysql/infoschema.go [hunk 21]
```diff
@@ -237,38 +228,99 @@ func (isi InfoSchemaImpl) GetColumns(conv *internal.Conv, table common.SchemaAnd
...
+
+	if col == "" || constraintType == "" {
+		conv.Unexpected("Got empty column or constraint type")
+		return nil
+	}
+
+	switch constraintType {
+	case "PRIMARY KEY":
+		*primaryKeys = append(*primaryKeys, col)
+	case "CHECK":
```
[the remark this answers is not in the document]
Comment (akashthawaitcc) [reply sources/mysql/infoschema.go:322]: This contraint type "CHECK" is coming from the MYSQL database when we list constraints.

File: webv2/api/schema.go [hunk 22]
```diff
@@ -485,7 +487,91 @@ func RestoreSecondaryIndex(w http.ResponseWriter, r *http.Request) {
...
+
+	convm := session.ConvWithMetadata{
+		SessionMetadata: sessionState.SessionMetadata,
+		Conv:            *sessionState.Conv,
+	}
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(convm)
+}
 
+func isCheckConstraintsNameExist(spcks []ddl.CheckConstraint, targetName string) bool {
```
[the remark this answers is not in the document]
Comment (taherkl) [reply webv2/api/schema.go]: Updated the function name to doesCheckConstraintNameExist

File: spanner/ddl/ast.go [hunk 23]
```diff
@@ -103,14 +103,16 @@ var PGSQL_TO_STANDARD_TYPE_TYPEMAP = map[string]string{
 
 // PGDialect keyword list
 // Assumption is that this list PGSQL dialect uses the same keywords
-var PGSQL_RESERVED_KEYWORD_LIST = []string{"ALL", "ANALYSE", "ANALYZE", "AND", "ANY", "ARRAY", "AS", "ASC", "ASYMMETRIC", "AUTHORIZATION", "BETWEEN", "BIGINT", "BINARY", "BIT", "BOOLEAN", "BOTH", "CASE", "CAST",
+var PGSQL_RESERVED_KEYWORD_LIST = []string{
```
[the remark this answers is not in the document]
Comment (taherkl) [reply spanner/ddl/ast.go]: Removed whitespaces

File: webv2/table/review_table_schema.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (taherkl) [reply webv2/table/review_table_schema.go:112]: The test is in "rename constraints column" under TestReviewTableSchema function

File: internal/mapping.go [hunk 24]
```diff
@@ -243,6 +243,19 @@ func ToSpannerIndexName(conv *Conv, srcIndexName string) string {
...
 
+// Note that the check constraints names in spanner have to be globally unique
+// (across the database). But in some source databases, such as MySQL,
+// they only have to be unique for a table. Hence we must map each source
+// constraint name to a unique spanner constraint name.
+func ToSpannerCheckConstraintName(conv *Conv, srcCheckConstraintName string) string {
+	return getSpannerValidName(conv, srcCheckConstraintName)
+}
+
+func GetSpannerValidExpression(cks []ddl.CheckConstraint) []ddl.CheckConstraint {
```
Comment (asthamohta) [inline internal/mapping.go:254]: nit: either use cc everywhere or ck

File: sources/common/toddl_test.go [hunk 25]
```diff
@@ -428,3 +428,34 @@ func Test_SchemaToSpannerSequenceHelper(t *testing.T) {
...
+
+	conv := internal.MakeConv()
+	srcSchema := []schema.CheckConstraint{
+		{
```
Comment (asthamohta) [inline sources/common/toddl_test.go:436]: Write a test where there is a name which isn't compatible

**Comment by asthamohta** [comment]
Where are the ADD and DELETE APIs for check constraints?

File: webv2/table/review_table_schema_test.go [hunk 26]
```diff
@@ -2060,6 +2095,59 @@ func TestReviewTableSchema(t *testing.T) {
 				},
 			},
 		},
+		{
```
Comment (asthamohta) [inline webv2/table/review_table_schema_test.go:2098]: Add a test for the sub string scenario

File: internal/mapping.go (hunk 24 above)
Comment (akashthawaitcc) [reply internal/mapping.go]: addressed everywhere in the code.

File: sources/common/toddl_test.go (hunk 25 above)
Comment (akashthawaitcc) [reply sources/common/toddl_test.go:436]: Added test case for the incompatible name.

File: webv2/api/schema_test.go [hunk 27]
```diff
@@ -2541,3 +2542,84 @@ func TestGetAutoGenMapMySQL(t *testing.T) {
 	}
 
 }
+func TestUpdateCheckConstraint(t *testing.T) {
+	sessionState := session.GetSessionState()
```
[the remark this answers is not in the document]
Comment (akashthawaitcc) [reply webv2/api/schema_test.go]: Merged all unit test cases.

File: webv2/table/review_table_schema_test.go (hunk 26 above)
Comment (akashthawaitcc) [reply webv2/table/review_table_schema_test.go:2098]: added test case for substring scenario

**Review by asthamohta — APPROVED** [review]

**Review by bharadwaj-aditya — APPROVED** [review]
LGTM
